File: ecobertura.ui/src/ecobertura/ui/launching/LaunchShortcutRetriever.java
package ecobertura.ui.launching;

import java.util.logging.Level;
import java.util.logging.Logger;

import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IConfigurationElement;
import org.eclipse.core.runtime.IExtensionPoint;
import org.eclipse.core.runtime.Platform;
import org.eclipse.debug.ui.IDebugUIConstants;
import org.eclipse.debug.ui.ILaunchShortcut;

public class LaunchShortcutRetriever {
	private static final Logger logger = Logger.getLogger("ecobertura.ui.launching"); //$NON-NLS-1$
	
	private String shortcutId;
	
	static ILaunchShortcut fromId(String shortcutId) {
		return new LaunchShortcutRetriever(shortcutId).getShortcut();
	}
	
	private LaunchShortcutRetriever(String shortcutId) {
		this.shortcutId = shortcutId;
	}

	private ILaunchShortcut getShortcut() {
		IExtensionPoint launchShortcuts = Platform.getExtensionRegistry().getExtensionPoint(
				IDebugUIConstants.PLUGIN_ID, IDebugUIConstants.EXTENSION_POINT_LAUNCH_SHORTCUTS);
		return getShortcutFromExtensionPoint(launchShortcuts);
	}

	private ILaunchShortcut getShortcutFromExtensionPoint(IExtensionPoint launchShortcuts) {
		for (IConfigurationElement config : launchShortcuts.getConfigurationElements()) {
			if (shortcutId.equals(config.getAttribute("id"))) { //$NON-NLS-1$
				return getLaunchShortcutFromConfigurationElement(config);
			}
		}
		logger.warning(String.format("No standard launch shortcut named '%s' found.", shortcutId)); //$NON-NLS-1$
		return null;
	}

	private ILaunchShortcut getLaunchShortcutFromConfigurationElement(
			IConfigurationElement config) {
		try {
			return (ILaunchShortcut) config.createExecutableExtension("class"); //$NON-NLS-1$
			
		} catch (CoreException e) {
			logger.log(Level.WARNING, "unable to create launch shortcut object", e); //$NON-NLS-1$
			return null;
		}
	}
}

File: ecobertura.ui/src/main/scala/ecobertura/ui/launching/LaunchShortcutRetriever.scala
/*
 * This file is part of eCobertura.
 * 
 * Copyright (c) 2009, 2010 Joachim Hofer
 * All rights reserved.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
package ecobertura.ui.launching

import java.util.logging._

import org.eclipse.core.runtime._
import org.eclipse.debug.ui._

object LaunchShortcutRetriever {
	private val logger = Logger getLogger "ecobertura.ui.launching" //$NON-NLS-1$
	
	def fromId(shortcutId: String): ILaunchShortcut = {
		new LaunchShortcutRetriever(shortcutId).getShortcut
	}
}
class LaunchShortcutRetriever(shortcutId: String) {
	import LaunchShortcutRetriever._
	
	def getShortcut = {
		def getShortcutFromExtensionPoint(launchShortcuts: IExtensionPoint) = {
			val shortcut = launchShortcuts.getConfigurationElements.find(
					config => shortcutId == config.getAttribute("id"))
			shortcut match {
				case Some(foundShortcutConfig) => 
					getLaunchShortcutFromConfigurationElement(foundShortcutConfig)
				case None => {
					logger warning String.format("No standard launch shortcut named '%s' found.", shortcutId) //$NON-NLS-1$
					null
				}
			}
		}
		
		def getLaunchShortcutFromConfigurationElement(config: IConfigurationElement) = {
			try {
				config.createExecutableExtension("class").asInstanceOf[ILaunchShortcut] //$NON-NLS-1$
			} catch {
				case e: CoreException => {
					logger.log(Level.WARNING, "unable to create launch shortcut object", e) //$NON-NLS-1$
					null
				}
			}
		}
		
		val launchShortcuts = Platform.getExtensionRegistry.getExtensionPoint(
				IDebugUIConstants.PLUGIN_ID, IDebugUIConstants.EXTENSION_POINT_LAUNCH_SHORTCUTS)
		getShortcutFromExtensionPoint(launchShortcuts)
	}
}
